### PR TITLE
EQL: avoid attempting PIT close on PIT open failure

### PR DESCRIPTION
**Pull request by luigidellaquila**
When an error happens during a PIT query, a PIT close is triggered.
In some specific circumstances, the PIT could be open at all (eg. if the failure was on openPIT), so the close operation could be invalid and lead to further errors. An example stack trace of this situation is:

```
java.lang.IllegalStateException: Channel is already closed
	at org.elasticsearch.rest.RestController$ResourceHandlingHttpChannel.close(RestController.java:670)
	at org.elasticsearch.rest.RestController$ResourceHandlingHttpChannel.sendResponse(RestController.java:663)
	at org.elasticsearch.xpack.eql.plugin.RestEqlSearchAction$1.onFailure(RestEqlSearchAction.java:94)
	at org.elasticsearch.rest.action.RestCancellableNodeClient$1.onFailure(RestCancellableNodeClient.java:96)
	at org.elasticsearch.client.internal.node.NodeClient$ActionResponseTaskListener.onFailure(NodeClient.java:185)
	at org.elasticsearch.tasks.TaskManager$1.onFailure(TaskManager.java:185)
	at org.elasticsearch.action.support.ContextPreservingActionListener.onFailure(ContextPreservingActionListener.java:38)
	at org.elasticsearch.action.ActionListener$Delegating.onFailure(ActionListener.java:92)
	at org.elasticsearch.xpack.ql.plugin.TransportActionUtils.lambda$executeRequestWithRetryAttempt$0(TransportActionUtils.java:71)
	at org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:170)
	at org.elasticsearch.xpack.eql.execution.PlanExecutor.lambda$eql$0(PlanExecutor.java:80)
	at org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:170)
	at org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:170)
	at org.elasticsearch.action.ActionListener$Delegating.onFailure(ActionListener.java:92)
	at org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:170)
	at org.elasticsearch.client.internal.node.NodeClient.doExecute(NodeClient.java:97)
	at org.elasticsearch.client.internal.support.AbstractClient.execute(AbstractClient.java:380)
	at org.elasticsearch.client.internal.FilterClient.doExecute(FilterClient.java:57)
	at org.elasticsearch.client.internal.ParentTaskAssigningClient.doExecute(ParentTaskAssigningClient.java:55)
	at org.elasticsearch.client.internal.support.AbstractClient.execute(AbstractClient.java:380)
	at org.elasticsearch.xpack.eql.execution.search.PITAwareQueryClient.close(PITAwareQueryClient.java:142)
	at org.elasticsearch.xpack.eql.execution.sequence.TumblingWindow.lambda$execute$1(TumblingWindow.java:129)
	at org.elasticsearch.action.ActionListener$RunAfterActionListener.onFailure(ActionListener.java:378)
	at org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:170)
	at org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:170)
	at org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:170)
	at org.elasticsearch.client.internal.node.NodeClient.doExecute(NodeClient.java:97)
	at org.elasticsearch.client.internal.support.AbstractClient.execute(AbstractClient.java:380)
	at org.elasticsearch.client.internal.FilterClient.doExecute(FilterClient.java:57)
	at org.elasticsearch.client.internal.ParentTaskAssigningClient.doExecute(ParentTaskAssigningClient.java:55)
	at org.elasticsearch.client.internal.support.AbstractClient.execute(AbstractClient.java:380)
	at org.elasticsearch.xpack.eql.execution.search.PITAwareQueryClient.openPIT(PITAwareQueryClient.java:134)
	at org.elasticsearch.xpack.eql.execution.search.PITAwareQueryClient.search(PITAwareQueryClient.java:59)
	at org.elasticsearch.xpack.eql.execution.search.BasicQueryClient.query(BasicQueryClient.java:64)
	at org.elasticsearch.xpack.eql.execution.sequence.TumblingWindow.advance(TumblingWindow.java:192)
	at org.elasticsearch.xpack.eql.execution.sequence.TumblingWindow.tumbleWindow(TumblingWindow.java:167)
	at org.elasticsearch.xpack.eql.execution.sequence.TumblingWindow.execute(TumblingWindow.java:127)
	at org.elasticsearch.xpack.eql.plan.physical.SequenceExec.execute(SequenceExec.java:122)
	at org.elasticsearch.xpack.eql.session.EqlSession.lambda$eql$0(EqlSession.java:85)
	at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.xpack.ql.util.ActionListeners.lambda$map$0(ActionListeners.java:24)
	at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.xpack.ql.util.ActionListeners.lambda$map$0(ActionListeners.java:24)
	at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.xpack.ql.util.ActionListeners.lambda$map$0(ActionListeners.java:24)
	at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.xpack.ql.util.ActionListeners.lambda$map$0(ActionListeners.java:24)
	at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.xpack.ql.index.IndexResolver.lambda$resolveAsMergedMapping$5(IndexResolver.java:367)
	at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.client.internal.node.NodeClient$ActionResponseTaskListener.onResponse(NodeClient.java:175)
	at org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:176)
	at org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:170)
	at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:31)
	at org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.lambda$applyInternal$2(SecurityActionFilter.java:165)
	at org.elasticsearch.action.ActionListener$DelegatingFailureActionListener.onResponse(ActionListener.java:245)
	at org.elasticsearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:47)
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:773)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
	Suppressed: java.lang.IllegalArgumentException: reader id must be specified
		at org.elasticsearch.action.search.ClosePointInTimeRequest.validate(ClosePointInTimeRequest.java:44)
		at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:43)
		at org.elasticsearch.tasks.TaskManager.registerAndExecute(TaskManager.java:170)
		at org.elasticsearch.client.internal.node.NodeClient.executeLocally(NodeClient.java:113)
		at org.elasticsearch.client.internal.node.NodeClient.doExecute(NodeClient.java:91)
		... 51 more
```

This PR adds a specific check, so that the PIT close is not executed if the PIT id is not valid (ie. there is no open PIT)

### CODE REVIEW

**Review by costin — APPROVED** [review]
LGTM

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sequence/PITFailureTests.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+import static org.elasticsearch.xpack.eql.plugin.EqlPlugin.CIRCUIT_BREAKER_LIMIT;
+import static org.elasticsearch.xpack.eql.plugin.EqlPlugin.CIRCUIT_BREAKER_NAME;
+import static org.elasticsearch.xpack.eql.plugin.EqlPlugin.CIRCUIT_BREAKER_OVERHEAD;
+
+public class PITFailureTests extends ESTestCase {
+
+    public static final String PIT_EXCEPTION_MESSAGE = "test - PIT open did not succeed";
+    private final List<HitExtractor> keyExtractors = emptyList();
+    private final HitExtractor tsExtractor = TimestampExtractor.INSTANCE;
+    private final HitExtractor implicitTbExtractor = ImplicitTiebreakerHitExtractor.INSTANCE;
```
Comment (bpintea) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sequence/PITFailureTests.java:75]: Do these need to be declared vs just use the instances where needed?

**Review by bpintea — APPROVED** [review]
Lgtm.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sequence/PITFailureTests.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+    public void testHandlingPitFailure() {
+        int sequenceFiltersCount = randomIntBetween(1, 5);
+        List<BreakerSettings> eqlBreakerSettings = Collections.singletonList(
```
Comment (Luegg) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sequence/PITFailureTests.java:79]: Why is the CircuitBreaker relevant for this test? Wouldn't one of the `NoopCircuitBreaker`s suffice?

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sequence/PITFailureTests.java [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+                new EqlFunctionRegistry(),
+                new Verifier(new Metrics()),
+                new Optimizer(),
+                new Planner(),
+                eqlCircuitBreaker
+            );
+            QueryClient eqlClient = new PITAwareQueryClient(eqlSession);
+            List<Criterion<BoxedQueryRequest>> criteria = new ArrayList<>();
+
+            for (int i = 0; i < sequenceFiltersCount; i++) {
```
Comment (Luegg) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sequence/PITFailureTests.java:139]: Is it relevant for this error case that there might be multiple criterions? Doesn't the query fail with the first request anyway?

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sequence/PITFailureTests.java [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+                    )
+                );
+            }
+
+            SequenceMatcher matcher = new SequenceMatcher(sequenceFiltersCount, false, TimeValue.MINUS_ONE, null, eqlCircuitBreaker);
+            TumblingWindow window = new TumblingWindow(eqlClient, criteria, null, matcher);
+            window.execute(
+                wrap(
+                    p -> { fail("Search succeeded despite PIT failure"); },
+                    ex -> { assertEquals(PIT_EXCEPTION_MESSAGE, ex.getMessage()); }
```
Comment (Luegg) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sequence/PITFailureTests.java:164]: `PIT_EXCEPTION_MESSAGE` can be inlined. Why the redirection?

**Review by Luegg — APPROVED** [review]
Generally LGTM but I would boil the test a bit more down to the essentials if possible.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sequence/PITFailureTests.java (hunk 4 above)
Comment (luigidellaquila) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sequence/PITFailureTests.java:134]: PIT_EXCEPTION_MESSAGE is used in two different places, that's why it's defined as a constant

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

**Review by astefan — APPROVED** [review]
LGTM. Thank you.

**Pull request merged**